arroyo-state/src/parquet.rs

```diff
@@ -378,48 +438,148 @@ impl BackingStore for ParquetBackend {
+                tuples_in.extend(tuples);
+            }
+        }
+
+        // do the compaction
+        let compactor: Compactor = Compactor::for_table_type(table_descriptor.table_type());
+        let tuples_length = tuples_in.len();
+        let tuples_out = compactor.compact_tuples(tuples_in);
+
+        if tuples_length != tuples_out.len() {
+            info!(
+                "Compaction summary for operator {}, table {}, task {}: {} tuples in, {} tuples out, {} tuples compacted",
+                task.operator_id,
+                table_char,
+                task.task_index,
+                tuples_length,
+                tuples_out.len(),
+                tuples_length - tuples_out.len()
+            );
+        }
+
+        let task_index = task.task_index.to_string();
+        let label_values = [
+            task.operator_id.as_str(),
+            task.job_id.as_str(),
+            task_index.as_str(),
+        ];
+
+        COMPACTION_TUPLES_IN_GAUGE
+            .with_label_values(&label_values)
+            .set(tuples_length as f64);
+        COMPACTION_TUPLES_OUT_GAUGE
+            .with_label_values(&label_values)
+            .set(tuples_out.len() as f64);
+
+        let mut parquet_writer =
+            ParquetCompressedFileWriter::new(table_char, new_min_epoch, task.clone());
+
+        for tuple in tuples_out {
+            parquet_writer.write(
+                tuple.key_hash,
+                tuple.timestamp,
+                tuple.key,
+                tuple.value,
+                tuple.operation,
+            );
+        }
+
+        let p = parquet_writer
+            .flush(&storage_client)
+            .await
+            .expect("Failed to flush parquet writer");
+        p
+    }
+
     async fn compact_operator(
+        parallelism: usize,
         job_id: String,
-        operator: String,
+        operator_id: String,
         old_min_epoch: u32,
         new_min_epoch: u32,
-    ) -> anyhow::Result<String> {
-        let paths_to_keep: HashSet<String> =
-            Self::load_operator_metadata(&job_id, &operator, new_min_epoch)
+    ) -> Result<(String, Option<CompactionResult>)> {
+        let operator_checkpoint_metadata =
+            Self::load_operator_metadata(&job_id, &operator_id, new_min_epoch)
                 .await
-                .expect("expect new_min_epoch metadata to still be present")
-                .backend_data
-                .iter()
-                .map(|backend_data| {
-                    let Some(BackendData::ParquetStore(parquet_store)) = &backend_data.backend_data
-                    else {
-                        unreachable!("expect parquet backends")
-                    };
-                    parquet_store.file.clone()
-                })
-                .collect();
+                .expect("expect new_min_epoch metadata to still be present");
+
+        // these files will be compressed, so the worker will be able to drop them
```

Comment:
I think we should try to use "compacted" rather than "compressed" throughout, as compression is happening on a different axis.